### ISSUE TITLE
INFO: Tunnel is up and running but for mac osx 10.14/10.15 I cannot resolve host

### ISSUE THREAD

**Issue by pjcau** [adrienverge/openfortivpn#501]
Hi Guys, 

I tryed use openFortiVPN, downloaded from brew v1.11 because I have to configure a mac osx machine to CI. So I tunnel up with info verbose log, 

```
INFO:   Adding VPN nameservers...
DEBUG:  ns1 already present in /etc/resolv.conf.
INFO:   Tunnel is up and running.
```

But I cannot resolve it. 

I use from command in CLI

`sudo openfortivpn XXX.YYY.com:ZZZ -u name --trusted-cert xxxxxxxxxxxxxx -v
`

Can you help me?

**Comment by DimitriPapadopoulos** [comment]
What do you mean by the following?
> But I cannot resolve it.

It looks like _openfortivpn_ does work. Please show us an error message.

**Comment by pjcau** [comment]
I need to use a VPN to resolve a host for git project. if I write a host

`ping http://myhost
`

return a 

`ping : cannot resolve  http://myhost:unknown host
`

**Comment by mrbaseman** [comment]
try `ping myhost` instead of `ping http://myhost`

**Comment by pjcau** [comment]
it's the same cannot dipend by protocol or port. But When I forticlient from fortinet, it works.

**Comment by mrbaseman** [comment]
is `myhost` a fully qualified hostname (including the domain)? 

With _openfortivpn_ 1.11.0 the domain search string that is sent from the Fortigate is added to `/etc/resolv.conf`. However, if you already have a `search` line in your `/etc/resolv.conf` the content is not merged (I think that's the same as the official Forticlient does), see  #472.

A simple workaround could be conncting via IP address or adding an entry to `/etc/hosts`. 

Also, if you notice a different behavior between _openfortivpn_ and _Forticlient_, it's always an appreciated feedback so that we can improve our code.

**Comment by DimitriPapadopoulos** [comment]
@pjcau I do realize you might not want to give real away real hostnames, but we need more information. This might an error on your side on issue with _openfortivpn_. We would typically need:
* the fully qualified hostname of the hostname you're trying to resolve,
* the contents of `/etc/resolv.conf` before running a VPN client,
* the contents of `/etc/resolv.conf` after running _openfotivpn_,
* the contents of `/etc/resolv.conf` after running _FortiClient_.

It's OK though to **consistently** obfuscate all domain names in the data you send us. For example:
`internal.mycompany.com` → `internal.domain1.net`
`mycompany.com` → `domain1.net`
`otherdomain.com` → `domain2.net`

**Comment by pjcau** [comment]
Yes, it's fully qualified. On Mac OS X on /etc/resolved.conf there is this file

nameserver yyyyyyyy
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
nameserver xxxxxxx

**Comment by DimitriPapadopoulos** [comment]
@pjcau Please change `yyyyyyyy` and `xxxxxxx` into something that looks like a domain name, even if **consistently** obfuscated. Also provide the hostname you're trying to resolve and the associated fully qualified hostname, again **consistently** obfuscated.

**Comment by DimitriPapadopoulos** [comment]
Also is this `/etc/resolv.conf`before or after running a VPN client? Which VPN client?

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman Martin, I've just read that on macOS the real file is actually `/var/run/resolv.conf`, not `/etc/resolv.conf`. I don't have a Mac, does this mean anything to you?
See [How to force MacOS to regenerate /etc/resolv.conf file?](https://apple.stackexchange.com/questions/264165/how-to-force-macos-to-regenerate-etc-resolv-conf-file)

**Comment by DimitriPapadopoulos** [comment]
@pjcau Thanks a lot, this helps. Clearly _openfortivpn_  does not set the DNS server while_FortiClient_ does set it: `nameserver` remains set to `123.123.123.567` instead of changing to `123.123.456.890`.
So it's not an issue with the domain lookup, ie. finding the fully qualified hostname from the short hostname. It's really an issue with DNS resolution of the fully qualified hostname.

**Comment by pjcau** [comment]
In fact @DimitriPapadopoulos , 

I do this yesterday 

```
sudo rm /etc/resolv.conf
sudo ln -s /var/run/resolv.conf /etc/resolv.conf
```

but it's the same

**Comment by DimitriPapadopoulos** [comment]
Oh, I didn't mean to say that's a `/var/run/resolv.conf` vs. `/etc/resolv.conf` issue. It won't help in this case.

That was a more general question.

**Comment by pjcau** [comment]
@DimitriPapadopoulos I know if isn't this issue for` /var/run/resolv.conf` vs. ` /etc/resolv.conf`. But I add only a info👍

**Comment by blakkheimgw** [comment]
Hello,

It seems I'm running the same sort of issue since the upgrade of openfortivpn fro 1.9.0 to 1.11.0 on my OpenSUSE Tumbleweed.

I'm using NetworkManager to manage my openfortivpn connections and it used to work well. Since the upgrade, the /etc/resolv.conf isn't updated anymore on connecting. If I force the usage of static DNS address in the NM menu, they are pushed in the /etc/resolv.conf but if I let the automatic option, it is not updated.

Thanks.

**Comment by DimitriPapadopoulos** [comment]
In theory it should work since issue #472 has been solved.

I think it would help if you could send the **sanitized**/**obfuscated** output of `openfortivpn -v -v -v`. The part of interest for us is described in https://github.com/adrienverge/openfortivpn/issues/472#issuecomment-539929149.

It's the part that contains:
```
<?xml version='1.0' encoding='utf-8'?><sslvpn-tunnel ver='2' dtls='1' patch='1'> ... <dns domain='dns-suffix.example.com' /><dns ip='192.168.xx.xx' /><dns ip='8.8.8.8' /> ... </sslvpn-tunnel>
```

**Comment by DimitriPapadopoulos** [comment]
The above just to make sure that the DNS server information is received from the FortiGate device. Once we can rule out an issue with DNS server reception, we'll have to look into the code that modifies `/etc/resolv.conf` based on the above information.

@blakkheimgw While I'm pretty certain it's the same issue, I think it would help if you could open a different issue with the requested information:
* the contents of `/etc/resolv.conf` before running a VPN client,
* the contents of `/etc/resolv.conf` after running _openfortivpn_,
* the output of the `nslookup` command showing the error, if possible before/after VPN connexion.

**Comment by pjcau** [comment]
Hi @DimitriPapadopoulos , I attach some debug

```
DEBUG:  http_receive:
HTTP/1.1 200 OK
Date: Thu, 05 Dec 2019 13:43:04 GMT
Server: xxxxxxxx-xxxxx
Transfer-Encoding: chunked
Content-Type: text/xml
X-Frame-Options: SAMEORIGIN
Content-Security-Policy: frame-ancestors 'self'
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Strict-Transport-Security: max-age=31536000

100d
<?xml version='1.0' encoding='utf-8'?><sslvpn-tunnel ver='2' dtls='1' patch='1'><dtls-config heartbeat-interval='10' heartbeat-fail-count='10' heartbeat-idle-timeout='10' client-hello-timeout='10' /><tunnel-method value='ppp' /><tunnel-method value='tun' /><fos platform='FGT31D' major='6' minor='00' patch='6' build='0272' branch='0272' /><client-config save-password='off' keep-alive='off' auto-connect='off' /><ipv4><dns ip='123.123.123.123' /><assigned-addr ipv4='123.123.111.123' /><split-tunnel-info><addr ip='23.123.111.555' mask='255.255.255.255' /><addr ip='123.123.111.222' mask='255.255.255.255' /></split-tunnel-info></ipv4><idle-timeout val='7200' /><auth-timeout val='36000' /></sslvpn-tunnel>
0


DEBUG:  found dns server 123.123.123.123 in xml configDEBUG:  Establishing the tunnel
DEBUG:  ppp_path: /usr/sbin/pppd
DEBUG:  Switch to tunneling mode
DEBUG:  http_send:
GET /remote/sslvpn-tunnel HTTP/1.1
Host: sslvpn
Cookie: blablablablablablablablablablablablablablablablablablablablablabla


```

**Comment by mrbaseman** [comment]
on my Mac the dns server is properly added to /etc/resolv.conf. Do you have by incidence a line `set-dns = 0` in your config file?

**Comment by pjcau** [comment]
@mrbaseman on my Mac I haven't the dns server on to` /etc/resolv.conf`.
And also I cannot use the config file, I use a this syntax

`sudo openfortivpn XXX.YYY.com:ZZZ -u name --trusted-cert xxxxxxxxxxxxxx -v
`

**Comment by mrbaseman** [comment]
@blakkheimgw NetworkManager integration is something different. The applet creates an ad-hoc configuration which it passes to openfortivpn when you connect. 

We have changed the default behavior in 1.11.0 that openfortivpn doesn't instruct pppd to request the nameservers but uses the ones it receives in the xml config (if nothing else is specified in the config file or on the command line).

Perhaps it makes sense to report this to the [NM-Applet developers](https://gitlab.gnome.org/GNOME/NetworkManager-fortisslvpn)

**Comment by mrbaseman** [comment]
@pjcau can you try appending `--set-dns=1` to your command line, please?

**Comment by pjcau** [comment]
@mrbaseman Now it becomes 

`sudo openfortivpn XXX.YYY.com:ZZZ -u name --trusted-cert xxxxxxxxxxxxxx -v --set-dns=1
`
but it's the same stuff.

**Comment by pjcau** [comment]
@DimitriPapadopoulos @mrbaseman Can I help in other way?

**Comment by mrbaseman** [comment]
hmm... I'm a bit lost now. The same stuff means, that the nameserver sent in the xml config is already in /etc/resolv.conf? And if so, then it's expected behavior. 
If Forticlient adds a different one, that would be strange, but then it's worth trying the `--pppd-use-peerdns=1` option, maybe together with `--set-dns=0` to avoid a race between openfortivpn and pppd

**Comment by pjcau** [comment]
@mrbaseman So on the top on this file 

 `/etc/resolv.conf` 

correctly add he namespace

` nameserver 123.123.456.890`

**Comment by mrbaseman** [comment]
So, on your system, the built-in update mechanism for /etc/resolv.conf does not work for some reason that we haven't found out yet, but pppd's usepeerdns option does the right thing, correct?

**Comment by pjcau** [comment]
so 
with `sudo openfortivpn XXX.YYY.com:ZZZ -u name --trusted-cert xxxxxxxxxxxxxx -v ` add 

```
nameserver 123.123.456.890
nameserver 123.123.123.1
```

**Comment by mrbaseman** [comment]
Sorry, but I'm completely lost here. First, you say that openfortivpn [does not alter](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562123941) your resolv.conf and [now you say that two lines are added](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562228484).
Is it by mistake adding IP addresses from the routes that are pushed as DNS servers?  

The first nameserver entry is the IP which _Forticlient_ has added [in this post](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562123941). 

The IP address 123.123.123.123 of the DNS server which has been received [according to here](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562137517) does **not** appear here at all. Instead you tell us that these IP addresses are [the gateway and the router](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562228484) on this machine. I don't think that I understand what this means about your network topology. 

But anyhow, is this picture in some way **consistent**? And if yes, may I ask the question if you have the same IP ranges in use in the local network at your client as the ones used remotely by the Fortigate?

If the picture is **not** consistent, could you please clarify what you _would expect to happen_ and what _actually is happening_ when openfortivpn tells you `DEBUG:  found dns server 123.123.123.123 in xml config` [like here](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562137517).

**Comment by pjcau** [comment]
OK, I correct some post that I wrote before to understand better. So ip gateway is 123.123.123.123. I fix it. But the issue is that some time the switch from Forticlient and openfortivpn, there are some issue in fact now the situation became

Before
```
nameserver 123.123.123.123
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#

```

OpenFortiVPN

```
nameserver 123.123.456.890
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
nameserver 123.123.123.1
```

Forticlient

```
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
nameserver 123.123.123.123

```

**Comment by DimitriPapadopoulos** [comment]
@pjcau So if I understand you correctly, you would expect the DNS server *not* to change: `nameserver` is `123.123.123.123` before running a VPN client and again `123.123.123.123` after starting _FortiClient_. Is that correct?

This is consistent with the XML part of the debug output you had sent us: 
```
<?xml version='1.0' encoding='utf-8'?><sslvpn-tunnel ver='2' dtls='1' patch='1'>...<dns ip='123.123.123.123' /><assigned-addr ipv4='123.123.111.123' />...</sslvpn-tunnel>
```

But then where do _openfortivpn_ or _pppd_ get `123.123.123.1` and `123.123.456.890` from?

I understand you're obfuscating the debug output and the contents of `/etc/resolv.conf` to hide real IP addresses and domain names. That's OK, but you absolutely need to do that in a **consistent** way. Is that indeed the case? Please double-check and maybe edit all your previous posts to fix any errors.

In any case, I don't see in the source code of _openfortivpn_ attempts to modify `/etc/resolv.conf` to anything else than the value sent by the FortiGate appliance (in your case `123.123.123.123`). So I don't think _openfortivpn_ is responsible for changing the value of `nameserver` to `123.123.456.890`and `123.123.123.1`. Perhaps _pppd_ is responsible for that but that's beyond our control as far as I can see, caused by some local _pppd_ configuration.

**Comment by DimitriPapadopoulos** [comment]
By the way, a good way to obfuscate IPv4 addresses would be to change only the "[network part](https://docs.oracle.com/cd/E19683-01/806-4075/ipref-1/index.html#ipref-2)" of the IP addresses (in a **consistent** way) and keep the "[host part](https://docs.oracle.com/cd/E19683-01/806-4075/ipref-1/index.html#auto27)". Use a different but consistent "network part" for each different network.

For example change these addresses:
```
128.93.162.1
128.93.162.84
128.93.162.254
```
to:
```
203.0.162.1
203.0.162.84
203.0.162.254
```

**Comment by pjcau** [comment]
> @pjcau So if I understand you correctly, you would expect the DNS server **not** to change: `nameserver` is `123.123.123.123` before running a VPN client and again `123.123.123.123` after starting _FortiClient_. Is that correct?
> 
> This is consistent with the XML part of the debug output you had sent us:
> 
> ```
> <?xml version='1.0' encoding='utf-8'?><sslvpn-tunnel ver='2' dtls='1' patch='1'>...<dns ip='123.123.123.123' /><assigned-addr ipv4='123.123.111.123' />...</sslvpn-tunnel>
> ```
> 
> But then where do _openfortivpn_ or _pppd_ get `123.123.123.1` and `123.123.456.890` from?
> 
> I understand you're obfuscating the debug output and the contents of `/etc/resolv.conf` to hide real IP addresses and domain names. That's OK, but you absolutely need to do that in a **consistent** way. Is that indeed the case? Please double-check and maybe edit all your previous posts to fix any errors.
> 
> In any case, I don't see in the source code of _openfortivpn_ attempts to modify `/etc/resolv.conf` to anything else than the value sent by the FortiGate appliance (in your case `123.123.123.123`). So I don't think _openfortivpn_ is responsible for changing the value of `nameserver` to `123.123.456.890`and `123.123.123.1`. Perhaps _pppd_ is responsible for that but that's beyond our control as far as I can see, caused by some local _pppd_ configuration.

I fix the getaway with `123.123.123.123` . So now it like this [post](https://github.com/adrienverge/openfortivpn/issues/501#issuecomment-562478874)

**Comment by DimitriPapadopoulos** [comment]
OK, so when running _FortiClient_ you get:
```
nameserver 123.123.123.123
```
and when running _openfortivpn_ you get:
```
nameserver 123.123.456.890
nameserver 123.123.123.1
```

Just to make sure, in the _openfortivpn_ case did you mean to write `123.123.123.123`:
```
nameserver 123.123.456.890
nameserver 123.123.123.123
```
or is it really `123.123.123.1`?
```
nameserver 123.123.456.890
nameserver 123.123.123.1
```
If it were `123.123.123.1` I would be really puzzled and wouldn't be able to make any sense out of it. The only solution would be to ask you to send real IP addresses (at least in private) to rule out any IP address conversion problem.

**Comment by DimitriPapadopoulos** [comment]
In any case this line:
`nameserver 123.123.456.890`
takes precendence over that one:
`nameserver 123.123.123.1`

Now I don't know where this `123.123.456.890` comes from - can't be _openfortivpn_. It has to be from the _pppd_ configuration. If so it might be that the (_pppd_) configuration of your Mac is broken. What does IP address `123.123.456.890` correspond to? Is it a DNS server?
More generally could you perhaps clarify the exact role of these IP addresses?
```
123.123.123.1
123.123.123.123
123.123.456.890
```

**Comment by pjcau** [comment]
It's crazy !!!! Now it works. I restart the mac and work....after standby doesn' it...After I restart and i't work.

```
nameserver 123.123.18.192
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
nameserver 123.123.0.1

```

this file is in openfortivpn. I change the  "network part" of the IPs.

**Comment by DimitriPapadopoulos** [comment]
I understand the above file is `/etc/resolv.conf` after starting _openfortivpn_, is that correct? Again `123.123.18.192` takes precedence over `123.123.0.1` here. Unfortunately your post doesn't help because we cannot compare to all the additional information you have provided so far. I can nevertheless try to provide some guidance below.

Are `123.123.18.192` and `123.123.0.1` both DNS servers? What are these servers in more detail? What happens when you run the following commands (hopefully works on macOS)?
```
nslookup myhost 123.123.18.192 
nslookup myhost 123.123.0.1
```

Also not sure what _standby_ really means here. I understand there are two distinct states for macOS machines: _sleep_ mode or _hibernate_ mode. The _standby_ mode really is _sleep_ followed by _hibernate_ after a delay. What you need to know is whether the machine was in _sleep_ mode (probably works well) or _hibernate_ (chances are the network setup is reset which could explain some of the trouble you have been experiencing).

Finally does it work now or not? If it always does then can you please close this issue? If not, see below.

To tell the truth, I'm afraid this has gotten too complicated to follow because you keep changing hostnames and IP addresses, so we only have partial and inconsistent information. We cannot help you this way. We need every single piece of information you send to be exact, complete, consistent, documented. At this point I suggest you close this issue and open a new one where you provide all the necessary information:
* contents of `/etc/resolv.conf` before running the VPN and after running _FortiClient_ and _openfortivpn_.
* consistent IPs and provide the role (corporate DNS server, ISP DNS server, local DNS server, router etc.) of every single IP address that appears in the issue you open
* consistent hostnames, use `myhost.mydomain` for fully qualified names and `myhost` for short names, and provide ythe IP addreses they are supposed to resolve to

If the above is too complicated and time-consuming, you still have the option to provide real IP addresses and hostnames.

**Comment by pjcau** [comment]
ok, we can do this. Moreover openfortivpn it works only after the restart, not in hibernate.

**Comment by justMaku** [comment]
I… have figured this out. The key to this issue is that macOS basically ignores /etc/resolv.conf (as stated in the file itself) and uses a different method of configuring it's DNS resolver. 

The intended way to add DNS servers to macOS is via a mechanism called SystemConfiguration that can be viewed and modified through either `scutil` command or a SystemConfiguration framwork in macOS SDK.

After dumping the system tree with this tool [dumpsystemconfiguration.py](https://gist.github.com/justMaku/b3597bdf633c648323127d9ac4ba14bf) before and after running official FortiClient I managed to understand what keys have to be added and changed to get the system wide DNS resolution to work.

By default openfortivpn only creates a `IPV4` entry in it's `State:/Network/Service/` namespace. To get DNS to work, you also need to create a `DNS` entry with a dictionary that contains an array of both nameservers retrieved from the VPN and the ones already existing in the system.

After that, you also need to modify the default `IPV4` entry to set `OverridePrimary` property to true, so it's the VPN service is the first one that system that system will query when doing DNS resolution.

I created a simple POC code that does that in Swift but it is entirely possible to also do that from pure C, I'm just not big enough of a masochist to attempt that. You can find my solution here: [main.swift](https://gist.github.com/justMaku/723f14cc03ed154478a5b5efebf28035)

**Comment by DimitriPapadopoulos** [comment]
@justMaku Hi, thank you for looking into this. May I suggest you create a new issue? This issue you have appended your post to is closed andxyou post will go unnoticed.

Also this issue a bit of a mess (which is one of the reasons why it has been closed in the first place) and might be unrelated.

**Comment by justMaku** [comment]
Sure! Added as #534, let me know if you need more details/testing, always happy to help.

**Comment by efrozo23** [comment]
I had the same problem, this command working for me 

openfortivpn -c config-file  -v --set-dns=0 --pppd-use-peerdns=1
